src/App.tsx

```diff
@@ -1,14 +1,69 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useState, useCallback } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
-
+import { getTodos } from './api';
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Todo } from './types/Todo';
+import { StatusFilter } from './helpers';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [status, setStatus] = useState(StatusFilter.ALL);
+  const [query, setQuery] = useState('');
+
+  useEffect(() => {
+    setIsLoading(true);
+
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+      })
+      .finally(() => setIsLoading(false))
+      .catch(error => new Error(`Error - todos was not loaded:', ${error}`));
+  }, []);
+
+  const handleStatus = useCallback((selectedStatus: StatusFilter) => {
+    setStatus(selectedStatus);
+  }, []);
+
+  const handleQuery = useCallback((search: string) => {
+    setQuery(search);
+  }, []);
+
+  const handleTodo = (todo: Todo) => {
+    setSelectedTodo(todo);
+  };
+
+  const clearSelectedTodo = () => {
+    setSelectedTodo(null);
+  };
+
+  const visibleTodos = todos.filter(todo => {
+    const normalizedTodo
+    = todo.title.toLowerCase().trim().includes(query.toLowerCase());
+
+    switch (status) {
+      case StatusFilter.ALL:
+        return normalizedTodo;
+      case StatusFilter.ACTIVE:
+        return normalizedTodo && !todo.completed;
+      case StatusFilter.COMPLETED:
+        return normalizedTodo && todo.completed;
+      default:
+        throw new Error('Unknown status selector');
+    }
+  });
```

Comment:
1. Let's move this logic to the helper and move it to `helpers` file.
2. Wrap `visibleTodos` value in useMemo